Remote GSM

efecarranza opened the pull request:
# Changelog

Add RemoteGSM Implementation
Abstract methods on GSM

miguelmtzinf commented on src/contracts/facilitators/gsm/RemoteGsm.sol:
```diff
@@ -0,0 +1,82 @@
+
+  /// @inheritdoc Gsm
+  function _getGhoOutstanding() internal view override returns (uint256) {
+    return MockCollector(_ghoVault).ghoOutstanding();
```
I think we better off creating a new escrow or vault contract that can be used to
- hold bridged GHO funds
- track accountability about used funds (analogous to level/capacity)
- pull/push functionality (analogous to mint/burn)

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -527,6 +553,14 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
   }
 
+  /**
+   * Returns the amount of GHO that has been sent out by the GSM
+   */
```
```suggestion
   * @dev Returns the total amount of GHO that has been spent
   * @param The amount of GHO that has been spent
   */
```
please amend the natspec docs of the rest of new functions

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -527,6 +553,14 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
+  /**
+   * Returns the amount of GHO that has been sent out by the GSM
+   */
+  function _getGhoOutstanding() internal view virtual returns (uint256) {
```
the name is not very clear to me. What about the following terms:
- getUsedGho: return the gho used by the contract
- getAvailableGho: return the gho limit that can be used by the contract
- useGho/restoreGho: transfers from/to the vault

miguelmtzinf replied:
we can also use `use`/`restore` and just `mint/burn` as we are minting and burning in practice (putting liquidity into circulation)

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -451,13 +452,38 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
     IERC20(UNDERLYING_ASSET).safeTransferFrom(originator, address(this), assetAmount);
 
-    IGhoToken(GHO_TOKEN).mint(address(this), grossAmount);
+    _handleGhoBought(grossAmount);
```
i d just rename to the term we agree on. (e.g. `useGho(grossAmount)`)
same with the others

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteVault.sol:
```diff
@@ -0,0 +1,59 @@
+import {AccessControl} from '@openzeppelin/contracts/access/AccessControl.sol';
+import {IGhoRemoteVault} from './interfaces/IGhoRemoteVault.sol';
+
+contract GhoRemoteVault is IGhoRemoteVault, AccessControl {
```
I d suggest to make it upgradeable

efecarranza commented:
@miguelmtzinf all comments have been addressed, thanks for the name suggestions, didn't know what to use!
GhoRemoteVault is a work in progress so any other feedback is more than welcome.

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteVault.sol:
```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import {IERC20} from '@aave/core-v3/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
+import {AccessControl} from '@openzeppelin/contracts/access/AccessControl.sol';
+import {VersionedInitializable} from '@aave/core-v3/contracts/protocol/libraries/aave-upgradeability/VersionedInitializable.sol';
+import {IGhoRemoteVault} from './interfaces/IGhoRemoteVault.sol';
+
+contract GhoRemoteVault is IGhoRemoteVault, AccessControl, VersionedInitializable {
```
```suggestion
import {IERC20} from '@aave/core-v3/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
import {VersionedInitializable} from '@aave/core-v3/contracts/protocol/libraries/aave-upgradeability/VersionedInitializable.sol';
import {IGhoRemoteVault} from './interfaces/IGhoRemoteVault.sol';

contract GhoRemoteVault is VersionedInitializable, IGhoRemoteVault {
```

miguelmtzinf replied:
let's add natspec docs here. Let's rename to `GhoReserve`, do you think is clearer?

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteVault.sol:
```diff
@@ -0,0 +1,80 @@
+   * @dev Throws if the caller does not have the FUNDS_ADMIN role
+   */
+  modifier onlyFundsAdmin() {
+    require(_onlyFundsAdmin(), 'ONLY_FUNDS_ADMIN');
```
I think we can just use a variable for `_fundsAdmin`

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteVault.sol:
```diff
@@ -0,0 +1,80 @@
+    IERC20(GHO_TOKEN).transferFrom(msg.sender, address(this), amount);
+  }
+
+  function bridgeGho(uint256 amount) external onlyFundsAdmin {
```
what's the purpose of this?

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteFacilitator.sol:
```diff
@@ -0,0 +1,39 @@
+    GHO_TOKEN = ghoAddress;
+  }
+
+  /**
+   * Mints specified amount of GHO to specified address
+   * @param receiver Address receiving GHO
+   * @param amount Amount of GHO to be minted
+   */
+  function mintAndBridge(address receiver, uint256 amount) external onlyOwner {
```
```suggestion
  /**
   * @notice Mint an amount of GHO to an address
   * @dev Only callable by the owner of the Facilitator.
   * @param account The address receiving GHO
   * @param amount The amount of GHO to be minted
   */
  function mint(address account, uint256 amount) external onlyOwner {
```
Please apply the convention reflected above across files and functions.
Also, `account` feels more aligned with `GhoToken` contract, and remove any mention to `bridge` as this will happen externally.

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteFacilitator.sol:
```diff
@@ -0,0 +1,39 @@
+   * @dev Constructor
+   * @param ghoAddress Address of GHO token on mainnet
+   */
+  constructor(address ghoAddress) {
```
let's pass the new owner. Same for GhoReserve

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,88 @@
+/**
+ * @title GhoReserve
+ * @author Aave/TokenLogic
+ * @notice GHO Remote Reserve. It provides withdraw/repay facilities to a GHO Stability Module in order to provide GHO liquidity on a remote chain.
```
I suggest to explain what is the purpose of this contract without mention to GSM here as it's totally decoupled.

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoRemoteFacilitator.sol:
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import {IERC20} from '@aave/core-v3/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
```
not needed it seems. Please check imports across files

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,88 @@
+ * @notice GHO Remote Reserve. It provides withdraw/repay facilities to a GHO Stability Module in order to provide GHO liquidity on a remote chain.
+ * @dev To be covered by a proxy contract.
+ */
+contract GhoReserve is IGhoReserve, Ownable, VersionedInitializable {
```
interface is usually put at the end, per Aave's codebase convention

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,88 @@
+    GHO_TOKEN = ghoAddress;
+  }
+
+  /**
+   * @notice GhoReserve initializer
+   * @param admin The address of the default admin role
+   */
+  function initialize(address admin) external initializer {
```
```suggestion
  /**
   * @dev Initializer
   * @param owner The address of the owner
   */
  function initialize(address owner) external initializer {
```
or an alternative like `newOwner` in case of shadowing

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  /// @inheritdoc IGhoReserve
+  function rescueToken(address token, address to, uint256 amount) external onlyOwner {
```
what's the purpose of this function? emergency rescue in case of mistakenly transfer of funds OR to pull unallocated GHO funds for mgmt purposes

efecarranza replied:
Multipurpose in my mind, both repay/send GHO back but also rescue. Can also rename and have it only transfer GHO.

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @title GhoReserve
+ * @author Aave/TokenLogic
+ * @notice GHO Remote Reserve. It provides withdraw/return facilities to approved contracts. Maximum withdraw capacity
+ * is specified per approved contract.
```
```suggestion
 * @notice It allows approved entities to withdraw and return GHO funds, with a defined maximum withdrawal capacity per entity.
```

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+  /// @inheritdoc IGhoReserve
+  address public immutable GHO_TOKEN;
+
+  /// @dev Mapping to keep track of GHO withdrawn by an address and capacity
```
```suggestion
  // Map of entities and their assigned capacity and amount of GHO used
```

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+
+  /**
+   * @dev Constructor
+   * @param initialOwner Address of the initial owner of the contract
```
```suggestion
   * @param initialOwner The address of the initial owner of the contract
```
per aave codebase convention. I suggest to use the same across functions

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+    require(ghoAddress != address(0), 'ZERO_ADDRESS_NOT_VALID');
+
+    _transferOwnership(initialOwner);
+
```
this break line is not needed imo

miguelmtzinf commented on src/contracts/facilitators/gsm/OwnableFacilitator.sol:
```diff
@@ -0,0 +1,60 @@
+ * @author Aave/TokenLogic
+ * @notice GHO Facilitator used to directly mint GHO to a given address.
+ */
+contract OwnableFacilitator is Ownable, VersionedInitializable, IOwnableFacilitator {
```
probably no need to make it upgradeable, as it's super-simple and easy to replace at GhoToken level. wdyt

efecarranza replied:
yeah, i didn't want to make it upgradeable to be honest but was just following what i had seen the most heh.

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -13,6 +13,7 @@ import {IGhoToken} from '../../gho/interfaces/IGhoToken.sol';
 import {IGsmPriceStrategy} from './priceStrategy/interfaces/IGsmPriceStrategy.sol';
 import {IGsmFeeStrategy} from './feeStrategy/interfaces/IGsmFeeStrategy.sol';
 import {IGsm} from './interfaces/IGsm.sol';
+import {IGhoReserve} from './interfaces/IGhoReserve.sol';
```
```suggestion
import {IGhoReserve} from './interfaces/IGhoReserve.sol';
import {IGsm} from './interfaces/IGsm.sol';
```
so it's `IGsm` at the end

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -107,17 +109,20 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
    * @param admin The address of the default admin role
    * @param ghoTreasury The address of the GHO treasury
    * @param exposureCap Maximum amount of user-supplied underlying asset in GSM
+   * @param ghoReserve Address of the GHO reserve to draw tokens from
```
```suggestion
   * @param ghoReserve The address of the GHO reserve to use tokens from
```

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -558,6 +600,21 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
   }
 
+  /**
+   * @dev Updates address of GHO reserve
```
```suggestion
   * @dev Updates the address of GHO reserve
```

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -558,6 +600,21 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
+   * @dev Updates address of GHO reserve
+   * @param ghoReserve The address of the GHO reserve for the GSM
+   */
+  function _updateGhoReserve(address ghoReserve) internal {
```
```suggestion
  function _updateGhoReserve(address newGhoReserve) internal {
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGsm.sol:
```diff
@@ -369,4 +391,10 @@ interface IGsm is IAccessControl, IGhoFacilitator {
    * @return The bytes32 signature typehash for sellAssetWithSig
    */
   function SELL_ASSET_WITH_SIG_TYPEHASH() external pure returns (bytes32);
+
+  /**
+   * @notice Updates the GHO reserve address
+   * @param ghoReserve The new address of the reserve holding GHO
+   */
+  function updateGhoReserve(address ghoReserve) external;
```
```suggestion

  /**
   * @notice Updates the GHO reserve address
   * @dev It revokes the allowance to the old reserve and grants maximum allowance to the new one.
   * @param newGhoReserve The new address of the GHO reserve
   */
  function updateGhoReserve(address newGhoReserve) external;
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGsm.sol:
Please re-order new functions so they are aligned with Solidity guide conventions for [order layout](https://docs.soliditylang.org/en/latest/style-guide.html#order-of-layout)

miguelmtzinf replied:
@efecarranza

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGsm.sol:
```diff
@@ -26,6 +26,13 @@ interface IGsm is IAccessControl, IGhoFacilitator {
   );
 
+  /**
+   * @dev Emitted when the GSM's reserve is updated
```
```suggestion
   * @dev Emitted when the GHO reserve is updated
```

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -237,14 +241,15 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
-    (, uint256 ghoMinted) = IGhoToken(GHO_TOKEN).getFacilitatorBucket(address(this));
-    if (amount > ghoMinted) {
-      amount = ghoMinted;
+    uint256 ghoOutstanding = _getUsedGho();
```
rename to `usedGho` or `ghoUsed` so it's consistent across files

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGsm.sol:
```diff
@@ -297,6 +304,21 @@ interface IGsm is IAccessControl, IGhoFacilitator {
+   */
+  function getGhoReserve() external view returns (address);
+
+  /**
+   * Returns amount of used GHO by the GSM
+   */
```
```suggestion
  /**
   * @notice Returns the amount of GHO used by the GSM
   * @return The amount of GHO used
   */
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGsm.sol:
```diff
@@ -297,6 +304,21 @@ interface IGsm is IAccessControl, IGhoFacilitator {
    */
   function getIsSeized() external view returns (bool);
 
+  /**
+   * Returns the address of the GHO reserve
+   */
```
```suggestion
  /**
   * @notice Returns the address of the GHO reserve
   * @return The address of the GHO reserve
   */
```

miguelmtzinf commented on src/script/DeployGsmLaunch.s.sol:
```diff
@@ -14,6 +14,8 @@ import {FixedFeeStrategy} from '../contracts/facilitators/gsm/feeStrategy/FixedF
 import {GsmRegistry} from '../contracts/facilitators/gsm/misc/GsmRegistry.sol';
 import {OracleSwapFreezer} from '../contracts/facilitators/gsm/swapFreezer/OracleSwapFreezer.sol';
 
+address constant GHO_RESERVE = address(0);
```
let's deploy instead

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -527,6 +553,22 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
+  }
+
+  /**
+   * @dev Returns the available amount of GHO left to withdraw
```
this is wrong

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+  function getCapacity(address withdrawer) external view returns (uint256) {
+    return _ghoCapacity[withdrawer].capacity;
+  }
+
```
let's add a `getUsage` getter to return both values, akin to `GHO.getFaciltiatorBucket()`

miguelmtzinf commented on src/test/TestGhoBase.t.sol:
```diff
@@ -273,13 +286,23 @@ contract TestGhoBase is Test, Constants, Events {
+      address(GHO_RESERVE)
+    );
+
+    GHO_RESERVE.setWithdrawerCapacity(address(GHO_GSM), 100_000_000 ether);
```
use `DEFAULT_CAPACITY` instead

miguelmtzinf commented on src/test/TestGsm.t.sol:
```diff
@@ -1330,9 +1332,6 @@ contract TestGsm is TestGhoBase {
     assertEq(seizedAmount, DEFAULT_GSM_USDC_AMOUNT, 'Unexpected seized amount');
 
-    vm.expectRevert('FACILITATOR_BUCKET_LEVEL_NOT_ZERO');
-    GHO_TOKEN.removeFacilitator(address(GHO_GSM));
```
This was to validate that we cannot offboard the facilitator yet. The analogous would be an assert that usedGho > 0

miguelmtzinf commented on src/test/TestGsm4626.t.sol:
```diff
@@ -406,8 +407,8 @@ contract TestGsm4626 is TestGhoBase {
 
-    (, ghoLevel) = GHO_TOKEN.getFacilitatorBucket(address(GHO_GSM_4626));
-    assertEq(ghoLevel, DEFAULT_CAPACITY - 1e18, 'Unexpected GHO bucket level after buy');
+    uint256 bucketLevel = GHO_GSM_4626.getUsedGho();
```
same, let's rename to `usedGho`

miguelmtzinf commented on src/test/TestGsm4626.t.sol:
```diff
@@ -924,9 +926,6 @@ contract TestGsm4626 is TestGhoBase {
     assertEq(seizedAmount, DEFAULT_GSM_USDC_AMOUNT, 'Unexpected seized amount');
 
-    vm.expectRevert('FACILITATOR_BUCKET_LEVEL_NOT_ZERO');
-    GHO_TOKEN.removeFacilitator(address(GHO_GSM_4626));
```
same as above. let's assert usedGho > 0

miguelmtzinf commented on src/test/TestGsm4626Edge.t.sol:
```diff
@@ -196,8 +196,8 @@ contract TestGsm4626Edge is TestGhoBase {
-    gsm.initialize(address(this), TREASURY, DEFAULT_GSM_USDC_EXPOSURE - 1);
-    GHO_TOKEN.addFacilitator(address(gsm), 'GSM Modified Exposure Cap', DEFAULT_CAPACITY);
+    gsm.initialize(address(this), TREASURY, DEFAULT_GSM_USDC_EXPOSURE - 1, address(GHO_RESERVE));
+    GHO_RESERVE.setWithdrawerCapacity(address(gsm), 100_000_000 ether);
```
should we use `DEFAULT_CAPACITY` instead? It can be a good idea to rename or create new constant for `GHO_RESERVE_DEFAULT_LIMIT`

miguelmtzinf commented on src/test/TestGsmSwapEdge.t.sol:
```diff
@@ -19,8 +19,8 @@ contract TestGsmSwapEdge is TestGhoBase {
-    gsm.initialize(address(this), TREASURY, type(uint128).max);
-    GHO_TOKEN.addFacilitator(address(gsm), 'GSM TINY', type(uint128).max);
+    gsm.initialize(address(this), TREASURY, type(uint128).max, address(GHO_RESERVE));
+    GHO_RESERVE.setWithdrawerCapacity(address(gsm), 100_000_000 ether);
```
```suggestion
    GHO_RESERVE.setWithdrawerCapacity(address(gsm), type(uint128).max);
```

miguelmtzinf commented on src/test/TestGsmSwapEdge.t.sol:
```diff
@@ -76,9 +76,9 @@ contract TestGsmSwapEdge is TestGhoBase {
+    gsm.initialize(address(this), TREASURY, 100_000_000e18, address(GHO_RESERVE));
     gsm.updateFeeStrategy(address(newFeeStrategy));
-    GHO_TOKEN.addFacilitator(address(gsm), 'Test GSM', DEFAULT_CAPACITY);
+    GHO_RESERVE.setWithdrawerCapacity(address(gsm), 100_000_000 ether);
```
same

miguelmtzinf commented on src/test/TestGsmSwapEdge.t.sol:
```diff
@@ -110,9 +110,9 @@ contract TestGsmSwapEdge is TestGhoBase {
+    gsm.initialize(address(this), TREASURY, 100_000_000e18, address(GHO_RESERVE));
     gsm.updateFeeStrategy(address(newFeeStrategy));
-    GHO_TOKEN.addFacilitator(address(gsm), 'Test GSM', DEFAULT_CAPACITY);
+    GHO_RESERVE.setWithdrawerCapacity(address(gsm), 100_000_000 ether);
```
same

miguelmtzinf commented on src/test/TestGsmSwapFuzz.t.sol:
```diff
@@ -89,8 +89,9 @@ contract TestGsmSwapFuzz is TestGhoBase {
-    gsm.initialize(address(this), TREASURY, type(uint128).max);
-    GHO_TOKEN.addFacilitator(address(gsm), 'GSM TINY', type(uint128).max);
+    gsm.initialize(address(this), TREASURY, type(uint128).max, address(GHO_RESERVE));
+    GHO_RESERVE.setWithdrawerCapacity(address(gsm), type(uint256).max);
```
it's `uint128` right

miguelmtzinf commented on src/test/TestGsmSwapFuzz.t.sol:
```diff
@@ -264,6 +265,8 @@ contract TestGsmSwapFuzz is TestGhoBase {
       amount = type(uint128).max;
     }
 
+    deal(address(GHO_TOKEN), address(GHO_RESERVE), type(uint256).max);
```
let's move it up together with `setWithdrawerCapacity`

miguelmtzinf commented on src/test/TestOwnableFacilitator.t.sol:
```diff
@@ -0,0 +1,98 @@
+    facilitator.initialize(address(0));
+  }
+
+  function testMint() public {
```
let's create a fuzz version of it

miguelmtzinf commented on src/test/TestOwnableFacilitator.t.sol:
```diff
@@ -0,0 +1,98 @@
+
+  function testRevertMintIfMintIsTooHigh() public {
+    vm.expectRevert('FACILITATOR_BUCKET_CAPACITY_EXCEEDED');
+    OWNABLE_FACILITATOR.mint(address(this), 101_000_000 ether);
```
fetch current capacity and add 1 instead

miguelmtzinf commented on src/test/TestOwnableFacilitator.t.sol:
```diff
@@ -0,0 +1,98 @@
+    assertEq(level, amount);
+    assertEq(amount, ghoBalanceAfter);
+
+    GHO_TOKEN.transfer(address(OWNABLE_FACILITATOR), amount / 2);
```
create separate tests for transfer, with the following cases:
1. transfer with no funds
2. transfer with funds
3. transfer with funds where requested amount is less and more than holdings
4. some mints and burns and then transfer of funds

miguelmtzinf commented on src/test/TestOwnableFacilitator.t.sol:
```diff
@@ -0,0 +1,98 @@
+
+  function testRevertBurnIfNoBalance() public {
+    vm.expectRevert();
+    OWNABLE_FACILITATOR.burn(50_000 ether);
```
this can be 1, also add a case where the are X burnable and we burn less

miguelmtzinf commented on src/test/TestGsmFullFlow.t.sol:
```diff
@@ -0,0 +1,89 @@
+      address(USDC_TOKEN),
+      address(GHO_GSM_FIXED_PRICE_STRATEGY)
+    );
+    gsm.initialize(address(this), TREASURY, DEFAULT_GSM_USDC_EXPOSURE, address(reserve));
```
is there deployment helpers? better to use them for GhoReserve as well

miguelmtzinf commented on src/test/TestGhoReserve.t.sol:
For example, for transfer:
- transfer with no funds
- transfer with funds
- transfer with funds where requested amount is less and more than holdings
- some mints and burns and then transfer of funds

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,97 @@
+/**
+ * @title GhoReserve
+ * @author Aave/TokenLogic
+ * @notice @notice It allows approved entities to withdraw and return GHO funds, with a defined maximum withdrawal capacity per entity.
```
```suggestion
 * @notice It allows approved entities to withdraw and return GHO funds, with a defined maximum withdrawal capacity per entity.
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  /**
+   * Transfers GHO to a specified address
+   * @param to Address receiving GHO token
+   * @param amount Amount of token to transfer
```
```suggestion
   * Emitted when GHO is transferred from this contract
   * @param to The address receiving the GHO tokens
   * @param amount The amount of GHO tokens
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * @dev Emitted when the GHO limit for a given entity is updated
+   * @param entity Address that can use GHO
+   * @param limit Maximum limit
```
```suggestion
   * @param entity The address of the entity
   * @param limit The maximum usage limit of the entity
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+   * @notice Returns the address of the GHO token
+   * @return The address of GHO token contract
+   */
+  function GHO_TOKEN() external view returns (address);
```
move this down right before `getUsed`

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+   */
+  function GHO_TOKEN() external view returns (address);
+
+  /**
+   * @notice Accepts GHO to be repaied by caller
+   * @param amount The amount of GHO to return
+   */
+  function restore(uint256 amount) external;
+
+  /**
+   * @notice Allows allowed caller to use GHO from reserve
+   * @param amount The amount of GHO to use
+   */
+  function use(uint256 amount) external;
```
```suggestion
  /**
   * @notice Uses a specified amount of GHO from the reserve.
   * @dev Callable only by entities with sufficient usage limit.
   * @param amount The amount of GHO to use.
   */
  function use(uint256 amount) external;
 
  /**
   * @notice Restores a specified amount of GHO to the reserve.
   * @dev The entity must grant allowance in advance to enable the reserve to pull the funds.
   * @dev Only callable by approved reserve entities.
   * @param amount The amount of GHO to restore.
   */
  function restore(uint256 amount) external;
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+   */
+  function transfer(address to, uint256 amount) external;
+
+  /**
+   * Sets a given addresses' limit
+   * @dev Only callable by the owner of the GhoReserve.
+   * @param entity Address that can use GHO
+   * @param limit Maximum amount of GHO that can be used
+   */
+  function setEntityLimit(address entity, uint256 limit) external;
```
```suggestion
  /**
   * @notice Sets a usage limit for a specified entity.
   * @param entity The address of the entity
   * @param limit The maximum amount of GHO that can be used
   */
  function setLimit(address entity, uint256 limit) external;
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+  function use(uint256 amount) external;
+
+  /**
+   * Rescues an ERC20 token by sending to a specified address
+   * @param to Address receiving the GHO token
+   * @param amount Amount of ERC20 token to transfer
+   */
```
```suggestion
   * @notice Transfers a specified amount of GHO from the reserve
   * @param to The address receiving the GHO tokens
   * @param amount The amount of GHO to transfer
   */
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+   * @param to Address receiving GHO token
+   * @param amount Amount of token to transfer
+   */
+  event GhoTokenTransfered(address to, uint256 amount);
```
```suggestion
  event GhoTransferred(address indexed to, uint256 amount);
```
Add analogous events for `GhoUsed` and `GhoRestored`

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+interface IGhoReserve {
```
add natspec docs

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IOwnableFacilitator.sol:
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+interface IOwnableFacilitator {
```
add natspec docs

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IOwnableFacilitator.sol:
```diff
@@ -0,0 +1,25 @@
+  /**
+   * Burns specified amount of GHO
+   * @dev Only callable by the owner of the Facilitator.
+   * @param amount Amount of GHO to be burned
```
```suggestion
   * @param amount The amount of GHO to be burned
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IOwnableFacilitator.sol:
```diff
@@ -0,0 +1,25 @@
+  function mint(address account, uint256 amount) external;
+
+  /**
+   * Burns specified amount of GHO
```
```suggestion
   * @notice Burns an amount of GHO
```

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -258,6 +263,11 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
   }
 
+  /// @inheritdoc IGsm
+  function updateGhoReserve(address ghoReserve) external onlyRole(CONFIGURATOR_ROLE) {
+    _updateGhoReserve(ghoReserve);
```
```suggestion
  function updateGhoReserve(address newGhoReserve) external onlyRole(CONFIGURATOR_ROLE) {
    _updateGhoReserve(newGhoReserve);
```

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm4626.sol:
```diff
@@ -43,14 +44,14 @@ contract Gsm4626 is Gsm, IGsm4626 {
 
-    (, uint256 ghoMinted) = IGhoToken(GHO_TOKEN).getFacilitatorBucket(address(this));
-    (, uint256 deficit) = _getCurrentBacking(ghoMinted);
+    uint256 ghoUsed = _getUsedGho();
```
change to `usedGho` for consistency

miguelmtzinf commented on src/script/DeployGsmLaunch.s.sol:
```diff
@@ -49,6 +50,30 @@ contract DeployGsmLaunch is Script {
+      ghoReserveInitParams
+    );
+
+    GhoReserve ghoReserve = GhoReserve(address(ghoReserveImpl));
```
must be `ghoReserveProxy` oinstead right?

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -527,6 +553,23 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
   }
 
+  /**
+   * @dev Returns the total amount of GHO that has been spent
+   * @return The amount of GHO that has been spent
```
```suggestion
   * @dev Returns the amount of GHO currently used
   * @return The amount of GHO used
```

miguelmtzinf commented on src/contracts/facilitators/gsm/Gsm.sol:
```diff
@@ -527,6 +553,23 @@ contract Gsm is AccessControl, VersionedInitializable, EIP712, IGsm {
+  }
+
+  /**
+   * @dev Returns the available amount of GHO that can be withdrawn by the GSM
+   * from the GHO reserve.
+   * @return The amount of GHO that can be withdrawn
```
```suggestion
   * @dev Returns the maximum amount of GHO that can be used.
   * @return The usage limit of GHO
```

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,97 @@
+
+  /// Map of entities and their assigned capacity and amount of GHO used
+  mapping(address => GhoUsage) private _ghoUsage;
+
```
Wondering if we should add global variables for total limit and total used

efecarranza replied:
Would total limit be every time we set a limit +/- to the global variable? For used it's easy to keep track

miguelmtzinf replied:
yeah, would be keeping global accounting every time an entity updates its usage (limit or used). Not sure it's completely needed

miguelmtzinf commented on src/test/TestGhoReserve.t.sol:
```diff
@@ -0,0 +1,195 @@
+    assertEq(limit - used, capacity - repayAmount);
+  }
+
+  function testsetEntityLimit() public {
```
typo

miguelmtzinf commented on src/test/TestGhoReserve.t.sol:
```diff
@@ -0,0 +1,195 @@
+    GHO_RESERVE.use(100 ether);
+  }
+
+  function testUseGho() public {
```
update test names according to new functions' names

miguelmtzinf commented on src/test/TestGhoReserve.t.sol:
```diff
@@ -0,0 +1,195 @@
+    assertEq(GHO_TOKEN.balanceOf(address(reserve)), 0);
+  }
+
+  function _deployReserve() public returns (GhoReserve) {
```
deploy proxy, initialize and return proxy

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+
+  /// @inheritdoc IGhoReserve
+  function use(uint256 amount) external {
+    GhoUsage memory usage = _ghoUsage[msg.sender];
+    require(usage.limit >= usage.used + amount, 'LIMIT_REACHED');
+
+    _ghoUsage[msg.sender].used += uint128(amount);
```
```suggestion
    GhoUsage storage entity = _ghoUsage[msg.sender];
    require(entity.limit >= entity.used + amount, 'LIMIT_REACHED');

    entity.used += uint128(amount);
```

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  /// @inheritdoc IGhoReserve
+  function setEntityLimit(address entity, uint256 limit) external onlyOwner {
```
```suggestion
  function setLimit(address entity, uint256 limit) external onlyOwner {
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,109 @@
+ */
+interface IGhoReserve {
+  /**
+   * Struct representing GSM's maximum allowed GHO usage and amount used
```
```suggestion
   * @dev Struct data representing GHO usage.
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,109 @@
+  function GHO_TOKEN() external view returns (address);
+
+  /**
+   * @notice Returns amount of GHO used by a specified address
+   * @param entity Address of the contract that withdrew GHO from reserve
```
```suggestion
   * @notice Returns the amount of GHO used by a specified address
   * @param entity The address of the entity
   * @return The amount of GHO used
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,109 @@
+  function getUsed(address entity) external view returns (uint256);
+
+  /**
+   * @notice Returns limit of GHO and used amount for a given entity
+   * @param entity Address of the contract that can use GHO from reserve
+   * @return Limit of GHO that can be used
+   * @return Used amount
```
```suggestion
   * @notice Returns the usage data of a specified entity
   * @param entity The address of the entity
   * @return The usage limit
   * @return The amount of GHO used
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,109 @@
+  function getUsage(address entity) external view returns (uint256, uint256);
+
+  /**
+   * @notice Returns maximum amount of GHO that can be used by a specified address
+   * @param entity Address of the contract that uses GHO from reserve
```
```suggestion
   * @notice Returns the usage limit of a specified entity
   * @param entity The address of the entity
   * @return The usage limit
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,109 @@
+   * @param entity The address that can use GHO
+   * @param limit The maximum usage limit of the entity
+   */
+  event EntityLimitUpdated(address indexed entity, uint256 limit);
```
```suggestion
  event GhoLimitUpdated(address indexed entity, uint256 limit);
```

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,109 @@
+
+  /**
+   * @dev Emitted when the GHO limit for a given entity is updated
+   * @param entity The address that can use GHO
+   * @param limit The maximum usage limit of the entity
```
```suggestion
   * @param entity The address of the entity
   * @param limit The new usage limit
```

miguelmtzinf commented on src/contracts/facilitators/gsm/OwnableFacilitator.sol:
```diff
@@ -0,0 +1,40 @@
+
+  /**
+   * @dev Constructor
+   * @param initialOwner Address of the initial owner of the contract
+   * @param ghoAddress Address of GHO token on mainnet
```
```suggestion
   * @param initialOwner The address of the initial owner
   * @param ghoAddress The address of GHO token
```

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+  /// @inheritdoc IGhoReserve
+  function use(uint256 amount) external {
+    GhoUsage storage entity = _ghoUsage[msg.sender];
+    require(entity.limit >= entity.used + amount, 'LIMIT_REACHED');
```
```suggestion
    require(entity.limit > entity.used + amount, 'LIMIT_EXCEEDED');
```
it should be > right? because if 100 is allocated, the entity can use 100.

miguelmtzinf commented on src/contracts/facilitators/gsm/interfaces/IGhoReserve.sol:
```diff
@@ -0,0 +1,111 @@
+   * @param entity The address restoring the GHO tokens
+   * @param amount The amount of token restored
+   */
+  event GhoUsed(address indexed entity, uint256 amount);
```
cosmetics nit, but i d move this up so it's easier to read: used, restored and transferred in order

DhairyaSethi commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,99 @@
+    require(initialOwner != address(0), 'ZERO_ADDRESS_NOT_VALID');
+    require(ghoAddress != address(0), 'ZERO_ADDRESS_NOT_VALID');
+
+    _transferOwnership(initialOwner);
```
this is an upgradeable contract, shouldn't mix setting immutable's and storage vars in constructor here, simply _disableInitializer for locking the impl and having null owner there is fine

efecarranza commented:
@miguelmtzinf I've added enumerableSet.
@DhairyaSethi i'm getting the following error once I removed the ownable portion of the constructor:


```
Error (7576): Undeclared identifier.
  --> src/contracts/facilitators/gsm/GhoReserve.sol:34:5:
   |
34 |     lastInitializedRevision = type(uint256).max;
   |     ^^^^^^^^^^^^^^^^^^^^^^^

Error: Compilation failed
```

any idea why this might be happening? VersionedInitializable doesn't have a _disableInitializers() function

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,132 @@
+
+  /// @inheritdoc IGhoReserve
+  function removeEntity(address entity) external onlyOwner {
+    GhoUsage memory usage = _ghoUsage[entity];
+    require(usage.used == 0, 'CANNOT_REMOVE_ENTITY_WITH_BALANCE');
```
```suggestion
    require(_ghoUsage[entity].used == 0, 'ENTITY_GHO_USED_NOT_ZERO');
```
no need of intermediate variable

miguelmtzinf commented on src/contracts/facilitators/gsm/GhoReserve.sol:
```diff
@@ -0,0 +1,132 @@
+  function removeEntity(address entity) external onlyOwner {
+    GhoUsage memory usage = _ghoUsage[entity];
+    require(usage.used == 0, 'CANNOT_REMOVE_ENTITY_WITH_BALANCE');
+    entities.remove(entity);
```
```suggestion
    require(entities.remove(entity), 'ENTITY_NOT_REMOVED');
```
returns false if it was not removed (because it was not present in the set for example).
Please add proper tests for this case and the one above.

miguelmtzinf commented on src/test/TestGhoReserve.t.sol:
```diff
@@ -96,9 +94,72 @@ contract TestGhoReserve is TestGhoBase {
+    emit EntityAdded(alice);
+    GHO_RESERVE.addEntity(address(alice));
+
+    // Set already contains two entities from constructor
```
fetch dynamically before, instead of hardcoding please

miguelmtzinf commented:
Closing in favor of https://github.com/aave-dao/gho-origin/pull/1
